### PR TITLE
[Merged by Bors] - Frustum Culling (for Sprites)

### PR DESCRIPTION
**Pull request by Byteron**
This PR adds two systems to the sprite module that culls Sprites and AtlasSprites that are not within the camera's view.
This is achieved by removing / adding a new  `Viewable` Component dynamically.

Some of the render queries now use a `With<Viewable>` filter to only process the sprites that are actually on screen, which improves performance drastically for scene swith a large amount of sprites off-screen.

https://streamable.com/vvzh2u

This scene shows a map with a 320x320 tiles, with a grid size of 64p.
This is exactly 102400 Sprites in the entire scene.

Without this PR, this scene runs with 1 to 4 FPS.

With this PR..
.. at 720p, there are around 600 visible sprites and runs at ~215 FPS
.. at 1440p there are around 2000 visible sprites and runs at ~135 FPS

The Systems this PR adds take around 1.2ms (with 100K+ sprites in the scene)

Note:
This is only implemented for Sprites and AtlasTextureSprites.
There is no culling for 3D in this PR.

### CODE REVIEW

File: crates/bevy_sprite/src/frustum_culling.rs
Comment (malu) [inline crates/bevy_sprite/src/frustum_culling.rs:85]: If the sprite is larger than the window, it might accidentally get culled here (imagine say a large background image). Checking whether the window rect and the sprites rect intersect would be preferrable in my opinion and would guard against this edge case.

Comment (Byteron) [reply crates/bevy_sprite/src/frustum_culling.rs:82]: This is a very good point. I'll change it.

Comment (Byteron) [reply crates/bevy_sprite/src/frustum_culling.rs]: Done

File: crates/bevy_render/src/draw.rs
Comment (alice-i-cecile) [inline crates/bevy_render/src/draw.rs:61]: Doc strings on the difference between Transparent and Visible would be nice.

File: crates/bevy_sprite/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_sprite/src/lib.rs:52]: These are probably going to need an ordering dependency on hierarchy propagation systems, but maybe that's better left out of this PR.

**Review by alice-i-cecile — APPROVED** [review]
It would be nice to have a benchmark to spot regressions and toss in the patch notes, but I think this PR is good to go and would be a really nice easy win for 0.5.

**Comment by alice-i-cecile** [comment]
This will partially resolve #1333.

**Comment by cart** [comment]
On paper I'm sold, although short lived marker components shouldn't be allowed until after we add support for SparseSet component storage (which will happen before 0.5. the draft pr should be out in the next couple of days). I'll give this a full review after I wrap that up. The only change required to adapt to that will be flagging the `Visible` component as "sparse" when we add the RenderPlugin.

File: crates/bevy_render/src/draw.rs
Comment (Byteron) [reply crates/bevy_render/src/draw.rs:61]: what are doc strings here?
just normal comments or something special?

**Comment by Byteron** [comment]
^I broke something when fixing clippy things. Fixed that and squashed commits.
(Hence the force push)

**Comment by Byteron** [comment]
> It would be nice to have a benchmark to spot regressions and toss in the patch notes, but I think this PR is good to go and would be a really nice easy win for 0.5.

Where would I put such a bench? Into `benches//bench/bevy_sprite/`?

File: crates/bevy_render/src/draw.rs
Comment (tigregalis) [reply crates/bevy_render/src/draw.rs:61]: [Doc Comments](https://doc.rust-lang.org/rust-by-example/meta/doc.html#doc-comments)

Comment (Byteron) [reply crates/bevy_render/src/draw.rs:63]: Also done.

**Comment by Byteron** [comment]
> It would be nice to have a benchmark to spot regressions and toss in the patch notes, but I think this PR is good to go and would be a really nice easy win for 0.5.

#1503 shall serve as bench / example project for this working.

**Comment by mockersf** [comment]
With `Visible` and `Transparent` as unit struct, how can one spawn an invisible or non-transparent `SpriteBundle`?

**Comment by rparrett** [comment]
That's a great question. I also asked in discord about manually managing visibility (but in a post-spawning scenario).

**Comment by rparrett** [comment]
Sprite rotation is also currently unaccounted for.

**Comment by Byteron** [comment]
> That's a great question. I also asked in discord about manually managing visibility (but in a post-spawning scenario).
> 
> https://discord.com/channels/691052431525675048/743663924229963868/813505768069398529
> 
> It sounds like Byteron's going to be looking into using a separate `Culled` marker struct instead, leaving the old `Visible` struct as-is.

> With `Visible` and `Transparent` as unit struct that are parts of the bundle, how can one spawn an invisible or non-transparent `SpriteBundle`?

Helloooo.
I reimplemented this with using a separate `Frustum` marker component.
Which means the `Visible` component stays untouched for this PR. There is also no need to manually add this Frustum Component yourself, that's done entirely by bevy.

> Sprite rotation is also currently unaccounted for.

This is still uncounted for. I need to figure out the math first in order to do this.

**Comment by aevyrie** [comment]
> > That's a great question. I also asked in discord about manually managing visibility (but in a post-spawning scenario).
> > https://discord.com/channels/691052431525675048/743663924229963868/813505768069398529
> > It sounds like Byteron's going to be looking into using a separate `Culled` marker struct instead, leaving the old `Visible` struct as-is.
> 
> > With `Visible` and `Transparent` as unit struct that are parts of the bundle, how can one spawn an invisible or non-transparent `SpriteBundle`?
> 
> Helloooo.
> I reimplemented this with using a separate `Frustum` marker component.
> Which means the `Visible` component stays untouched for this PR. There is also no need to manually add this Frustum Component yourself, that's done entirely by bevy.
> 
> > Sprite rotation is also currently unaccounted for.
> 
> This is still uncounted for. I need to figure out the math first in order to do this.

Could we use something like `WithinFrustum` that makes it more clear what the marker does? The name `Frustum` doesn't make it immediately obvious whether entities with the struct are inside our outside the camera frustum.

**Comment by Byteron** [comment]
> Could we use something like `WithinFrustum` that makes it more clear what the marker does? The name `Frustum` doesn't make it immediately obvious whether entities with the struct are inside or outside the camera frustum.

As I understand the Frustum is the visible part. So it makes sense to me that entities with Frustum are visible, while entities without Frustum are not.
Though if you think this is not descriptive enough I can rename it for sure.

**Comment by aevyrie** [comment]
A frustum is a shape, it doesn't really make much sense when we look at other component names:

- entities are `Visible`
- entities are `Transparent`
- entities are `Frustum` (?)

The first two make it clear what functionality the components add. Hence the suggestion for `WIthinFrustum`.

File: crates/bevy_sprite/src/frustum_culling.rs
Comment (aevyrie) [inline crates/bevy_sprite/src/frustum_culling.rs:52]: I think you can convert this into a parallel iterator, which might give a boost if you expect to iterate over a large number of items.
```rust
.par_iter_mut(32).for_each(
    &pool,
    |(entity, drawable_transform, sprite)| {
        // Do stuff
);
```

File: crates/bevy_sprite/src/frustum_culling.rs
Comment (aevyrie) [inline crates/bevy_sprite/src/frustum_culling.rs:99]: This could also be turned into a parallel iterator.

**Comment by Byteron** [comment]
Hm. I see your point.
Though I don't quite like the double word naming.
`Observable` maybe? `Ocular`?

File: crates/bevy_sprite/src/frustum_culling.rs
Comment (aevyrie) [inline crates/bevy_sprite/src/frustum_culling.rs:17]: This function could be replaced with something that allows for sprite rotation. 

1. Get each of the 4 vertices that define the corners of the sprite by applying the transform to the corners.
2. For each line of infinite length that defines the edge of the screen, check if all vertices are on the outside of the line. E.g. check if all sprite vertex x values are less than the screen's x position.
3. If they are, end early. The sprite is not in the camera frustum.
4. If the condition is not met for all four screen edges, the sprite is within the camera frustum.

File: crates/bevy_render/src/draw.rs
Comment (aevyrie) [inline crates/bevy_render/src/draw.rs:68]: I already suggested some alternate, more descriptive names in the comments. 😄

**Comment by aevyrie** [comment]
> Hm. I see your point.
> Though I don't quite like the double word naming.
> `Observable` maybe? `Ocular`, `Viewable`?

Might need to leave this up to cart. Personally, I don't see a problem with multi-word names. Why trade clarity for one less capital letter?

File: crates/bevy_sprite/src/frustum_culling.rs
Comment (Byteron) [reply crates/bevy_sprite/src/frustum_culling.rs]: I got around this by just using a shere radius for both sprites and the camera. That means there are a few more sprites rendered than necessary, but it keeps the algorithm simple and still massively improves performance.

File: crates/bevy_sprite/src/frustum_culling.rs
Comment (Byteron) [reply crates/bevy_sprite/src/frustum_culling.rs]: done

File: crates/bevy_sprite/src/lib.rs
Comment (Byteron) [reply crates/bevy_sprite/src/lib.rs:52]: I think they also need to be ordered with the visible_entities system in camera I think.
IIRC they are both called on Update and that might lead to sprites popping up on the camera's borders a frame later. It's barely noticable but it *is* noticable.

**Comment by Byteron** [comment]
Updated after ECSv2 landed.
the `Viewable` component is now a `StorageType::SparseSet` as cart mentioned it should be earlier in this convo.
performance gains remain roughly the same.

The `StorageType::SparseSet` leads to slower iteration times in some rendering related systems (as they use `With<Viewable>`), so as soon as queries allow sparse set components to be in filters while still have fast iteration times there should be another small performance bump.

File: crates/bevy_render/src/draw.rs
Comment (JeanMertz) [inline crates/bevy_render/src/draw.rs:79]: Can't this be derived?

File: crates/bevy_render/src/pipeline/render_pipelines.rs
Comment (JeanMertz) [inline crates/bevy_render/src/pipeline/render_pipelines.rs:10]:
```suggestion
    With,
```

File: crates/bevy_render/src/shader/shader_defs.rs
Comment (JeanMertz) [inline crates/bevy_render/src/shader/shader_defs.rs:6]:
```suggestion
    With,
```

File: crates/bevy_sprite/src/frustum_culling.rs
Comment (JeanMertz) [inline crates/bevy_sprite/src/frustum_culling.rs:35]: Just making sure/highlighting if we're okay with unwrapping in Bevy-provided systems (as opposed to logging and exiting cleanly)?

File: crates/bevy_sprite/src/frustum_culling.rs
Comment (JeanMertz) [reply crates/bevy_sprite/src/frustum_culling.rs:87]: Given that this a regular iterator, I assume you backed out of this change @Byteron?

File: crates/bevy_sprite/src/frustum_culling.rs
Comment (JeanMertz) [inline crates/bevy_sprite/src/frustum_culling.rs:99]:
```suggestion
                            commands.insert(entity, Viewable);
```

File: crates/bevy_sprite/src/lib.rs
Comment (JeanMertz) [inline crates/bevy_sprite/src/lib.rs:59]: nit: since the module is called `frustum_culling`, it feels somewhat redundant to have the same name appear in the function. Perhaps `frustum::culling::{sprites, atlases}`?

**Review by JeanMertz — COMMENTED** [review]
Sweet! I left a few comments, and also agree with @aevyrie that I'd prefer `InsideFrustum` or some other term that's more clear than `Visible`, but I'm sure @cart has an opinion on this as well.

File: crates/bevy_render/src/pipeline/render_pipelines.rs
Comment (Byteron) [reply crates/bevy_render/src/pipeline/render_pipelines.rs:10]: that does not compile for me.

Comment (Byteron) [reply crates/bevy_render/src/pipeline/render_pipelines.rs:10]: image
seems to be the only option here.

File: crates/bevy_sprite/src/frustum_culling.rs
Comment (Byteron) [reply crates/bevy_sprite/src/frustum_culling.rs]: dealt with unwrap to be on the safe side.

File: crates/bevy_render/src/pipeline/render_pipelines.rs
Comment (Byteron) [reply crates/bevy_render/src/pipeline/render_pipelines.rs]: it's `bevy_ecs::query::With` now.

**Comment by rparrett** [comment]
I updated the example locally to non-uniformly scale the sprites
```rs
let scale = Vec3::new(0.2 + rng.gen::<f32>() * 4.8, 1.0, 1.0);
```
And I am seeing some popping-in and popping-out around the edges that I didn't notice before making that change.

File: crates/bevy_render/src/camera/visible_entities.rs
Comment (cart) [inline crates/bevy_render/src/camera/visible_entities.rs:207]: As-is this completely breaks all 3d rendering (and any custom non-sprite rendering) because `WithinFrustum` is never added to those entities. We should probably add another system that adds `WithinFrustum` to every  `Query<Entity, (With<MainPass>, Without<Sprite>, Without<TextureAtlasSprite>)>`

**Review by cart — COMMENTED** [review]
Having thought about this a bit, I don't think we should use marker components here:

1. This won't work for multi-camera situations (ex: render to texture, split screen, editors). Each camera needs its own visibility set, but the marker component effectively merges those sets together. 
2. As much as possible I'd like Bevy to treat components as "user defined state", which it derives internal state from. That feels "cleaner" to me. I'm not sure we'll always be able to do it, but so far we've been able to.

Instead I think it makes sense for each camera to store a list of entities that are within the frustum (or for the renderer to populate that on the fly). However this makes the "query filter optimizations" you added to the various system (ex: shader def and draw system) harder to do (which would in turn make the wins here smaller).

I think all of those issues go away if we adopt the "pipelined rendering" model we've been discussing in various places, where we extract world state each frame and construct draw state on the fly. That would lend itself nicely to "dynamic per camera visibilty lists"

Buuut that work hasn't even _started_ yet and won't land until well after 0.5. The implementation in this pr (while limited) works for the common case and gives us a nice win. And it doesn't outright break multiple cameras/views (it just makes them draw stuff they don't see).

I think if we can sort out a good solution to the broken 3d (and custom) rendering, we should merge this.

File: crates/bevy_render/src/camera/visible_entities.rs
Comment (Byteron) [reply crates/bevy_render/src/camera/visible_entities.rs:207]: I added the suggested system to frustum_culling.rs.
the 3d_scene example does run as it should.

Comment (cart) [reply crates/bevy_render/src/camera/visible_entities.rs:207]: Hmm this is a bit problematic because it _requires_ the SpritePlugin to draw anything. Thats a bit too hack-ey for my taste. Sprites should be a layer on top of rendering, not a required piece. Thinking on this now.

Comment (cart) [reply crates/bevy_render/src/camera/visible_entities.rs:207]: I think maybe switching to a `Without<OutsideFrustum>` filter instead might do the trick.

File: crates/bevy_sprite/src/lib.rs
Comment (cart) [reply crates/bevy_sprite/src/lib.rs:61]: Unfortunately ordering won't be enough because the component insert commands won't be applied until after the PostUpdate stage finishes. There are a number of systems in PostUpdate that filter on that component's existence.

I think we _should_ add the relevant dependencies to be as correct as possible (and to ensure transform updates are applied), but we'll still have a one frame lag to cull / uncull sprites unless we either:

1. Ensure _all_ systems that rely on the frustum component live in a stage _after_ PostUpdate. This isn't ideal.
2. Do frustum culling later, remove `With<WithinFrustum>` and accept that we'll do unnecessary work for those systems (ex: we only filter out unnecessary draw calls, not the cpu work like shader management). This also isn't ideal.
3.  Move all of the steps out of the ECS (something like the "pipelined rendering" we've been discussing lately). This is a lot of work.

I'm hesitant to do (1) as it feels short sighted. Given how much cpu work is a bottleneck right now (2) doesn't feel like its "worth it".

So imo our only real options are "don't include frustum culling in 0.5 and wait for pipelined rendering" or "accept one frame of cull lag"

**Comment by cart** [comment]
Just pushed the changes mentioned here: https://github.com/bevyengine/bevy/pull/1492#discussion_r600073824

This "inverts" the queries so that they search for `Without<OutsideFrustum>`. This allows non-sprite rendering to work by default without hacks (and notably without requiring SpritePlugin).

**Comment by cart** [comment]
I think there might be a bug with insertion commands, which @Byteron may have already encountered given that they included checks that shouldn't be necessary, which i tried removing and it resulted in a bundle insertion panic. Re-adding the check resolved the panic, but this feels like a nasty bevy_ecs bug that needs solving before 0.5. Investigating now.

**Comment by cart** [comment]
Found the issue. It was a bevy_ecs bug introduced in the "reliable change detection" pr. Simple fix.

File: crates/bevy_render/src/draw.rs
Comment (JeanMertz) [inline crates/bevy_render/src/draw.rs:70]: This comment needs updating with the recent name change (and inversion of logic).

File: crates/bevy_render/src/camera/visible_entities.rs
Comment (JeanMertz) [inline crates/bevy_render/src/camera/visible_entities.rs:208]: Double negatives 🤯

_(nothing we can do here, but always causes me to pause for a moment when reading logic like this)_

Comment (cart) [reply crates/bevy_render/src/camera/visible_entities.rs:208]: Yeah I like to avoid double negatives in apis whenever possible. But yeah the tradeoff here for the inverse is too much :)

**Comment by cart** [comment]
An issue with using `OutsideFrustum` instead of `InsideFrustum` is that it won't support multiple views/cameras correctly. The old impl "under culled" objects viewed by multiple cameras. The new impl will "over cull" objects, which is much worse.  We don't support that sort of thing officially, but I think we should provide an escape hatch in case someone hits this issue in a custom render pipeline. Something like `SpriteSettings { frustum_culling_enabled: false }`.

**Comment by cart** [comment]
We'll also want to switch to only culling using the "active camera" to support multiple cameras of the same "type" (ex: switching back and forth between two "2d" cameras from different perspectives.

**Comment by cart** [comment]
(working on these changes now)

**Comment by alice-i-cecile** [comment]
> We don't support that sort of thing officially, but I think we should provide an escape hatch in case someone hits this issue in a custom render pipeline. Something like SpriteSettings { frustum_culling_enabled: false }.

#1389 is quite mature FWIW. I'd prefer to see a real integration here, but that can wait.

**Comment by cart** [comment]
> #1389 is quite mature FWIW. I'd prefer to see a real integration here, but that can wait.

Integration with multiple views would involve a complete rewrite / moving away from marker components (as discussed above). That definitely needs to wait.

**Pull request closed without merging**
